File: front-end/Mobile/src/models/attendance/JustificationType.js
import BaseModel from '../BaseModel';

export default class JustificationType extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.justificationTypeId = data.justification_type_id || null;
    this.name = data.name || '';
    this.description = data.description || null;
    this.requiresAttachment = data.requires_attachment !== undefined ? data.requires_attachment : false;
    this.status = data.status !== undefined ? data.status : true;
  }

  static fromApi(data) {
    if (!data) return null;
    return new JustificationType(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      justification_type_id: this.justificationTypeId,
      name: this.name,
      description: this.description,
      requires_attachment: this.requiresAttachment,
      status: this.status,
    };
  }
}

File: back-end/05-ms-attendance/src/main/java/com/faceattend_edu/attendance_service/domain/model/JustificationType.java
package com.faceattend_edu.attendance_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.Objects;
import java.util.UUID;

@Getter @Setter @AllArgsConstructor @NoArgsConstructor
public class JustificationType {
    private Integer justificationTypeId;
    private Integer schoolId;
    private String name;
    private String description;
    private Boolean requiresAttachment;
    private Boolean status;
    private Instant createdAt;
    private Instant updatedAt;
    private Instant deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    public void validate() {
        if (isBlank(name)) throw new IllegalArgumentException("name is required");
        if (name.length() > 255) throw new IllegalArgumentException("name max 255");
        if (description != null && description.length() > 500) throw new IllegalArgumentException("description max 500");
        if (requiresAttachment == null) requiresAttachment = false;
        if (status == null) status = true;
    }
    public void touchCreated() {
        if (createdAt == null) createdAt = Instant.now();
        if (requiresAttachment == null) requiresAttachment = false;
        if (status == null) status = true;
        if (rowVersion == 0) rowVersion = 1L;
    }
    public void touchUpdated() { updatedAt = Instant.now(); rowVersion++; }
    private static boolean isBlank(String s){ return s==null || s.trim().isEmpty(); }
    @Override public boolean equals(Object o){ if(this==o) return true; if(!(o instanceof JustificationType)) return false; JustificationType that=(JustificationType)o; return Objects.equals(justificationTypeId, that.justificationTypeId); }
    @Override public int hashCode(){ return Objects.hash(justificationTypeId); }
}
